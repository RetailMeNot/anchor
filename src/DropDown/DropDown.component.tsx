// REACT
import * as React from 'react';
// VENDOR
import classNames from 'classnames';
import styled from 'styled-components';
import { fromEvent, Subscription } from 'rxjs';
import { rgba } from 'polished';
import { get } from 'lodash';
// COMPONENTS
// THEME
import { colors, variables, fonts, sizes } from '../theme';

interface DropDownProps {
  overlay: React.ReactElement<any> | Array<React.ReactElement<any>> | React.SFC | React.SFC[];
  trigger?: 'hover' | 'click'; // TODO: 'contextMenu';
  className?: string;
  children?: any;
  // Configuration
  ref?: any;
}

interface DropDownState {
  isHidden: boolean;
  height?: number;
  width?: number;
}

export const StyledDropDown = styled.div`
  font-family: ${fonts.fontFamily};
  position: relative;
  cursor: pointer;
  line-height: 1rem;
  display: inline-flex;
  justify-content: center;
  align-items: center;

  .drop-down-container {
    min-width: 10rem;
    position: absolute;
    left: 0;
    top: ${(props: any) => `${get(props, 'forwardedRef.current.clientHeight', 16)}px`};
    padding: ${sizes.padding.xs} 0;
    z-index: 1;
    background-color: ${colors.neutrals.white.base};
    box-shadow: 0 .5rem .5rem 0 ${rgba(colors.neutrals.charcoal.dark, 0.12)};
    border-radius: ${variables.borderRadius};
    border: 1px solid ${colors.neutrals.silver.dark};
  }

  .arrow-up {
    position: absolute;
    top: -.4rem;
    left: 1rem;
    width: 0;
    height: 0;
    border-left: .75rem solid transparent;
    border-right: .75rem solid transparent;
    border-bottom: .5rem solid white;
  }
`;

export class DropDown extends React.Component<DropDownProps> {
  static defaultProps = {
    trigger: 'hover',
  };
  // State
  state: DropDownState = { isHidden: true };
  // Instance Reference
  private readonly dropDownReference: React.RefObject<DropDown>;
  // Observables
  // private escapeKey$: Subscription;
  // private windowClick$: Subscription;
  // private elementClick$: Subscription;
  private mouseOver$: Subscription;
  private mouseOut$: Subscription;

  constructor(props: DropDownProps) {
    super(props);
    this.dropDownReference = React.createRef();
  }

  componentDidMount(): void {
    const { current: dropDown }: { current: any } = this.dropDownReference;
    const { clientHeight: height, clientWidth: width } = dropDown;
    this.setState({ ...this.state, height, width });
    /* Events */
    /*
    // Subscribe to all window clicks
    this.windowClick$ = fromEvent(window, 'click')
      .pipe(
        takeWhile(() => {
          return !this.state.isHidden;
        })
      );
    // Subscribe to keyboard escape
    this.escapeKey$ = fromEvent(document, 'keyup')
      .pipe(
        takeWhile(() => !this.state.isHidden),
        filter((keyEvent: KeyboardEvent) => keyEvent.key === 'Escape')
      );
    // Subscribe to element click/tap
    this.elementClick$ = fromEvent(dropDown, 'click');
    */

    // If hover is the trigger:
    // Subscribe to element mouseenter
    this.mouseOver$ = fromEvent(dropDown, 'mouseenter')
      .subscribe((a: any) => {
        this.setState({ isHidden: false });
      });
    // Subscribe to element mouseleave
    this.mouseOut$ = fromEvent(dropDown, 'mouseleave')
      .subscribe((a: any) => {
        this.setState({ isHidden: true });
      });

    /* Subs */
    // this.windowClick$.subscribe((a: any) => {
    //   console.log(1, a, this.dropDownReference);
    //   // this.setState({ isHidden: true });
    // });
    // this.elementClick$.subscribe((a: any) => {
    //   console.log(2, a, this.dropDownReference);
    //   // this.setState({ isHidden: true });
    // });

    // const clickEvents$ = zip(this.windowClick$, this.elementClick$)
    //   .subscribe((...a) => {
    //     console.log(6, a);
    //   });

    // this.escapeKey$.subscribe((a: any) => {
    //   this.setState({ isHidden: true });
    // });
  }

  // componentDidUpdate(prevProps: Readonly<DropDownProps>, prevState: Readonly<{}>, snapshot?: any): void {
  //
  // }

  componentWillUnmount(): void {
    this.mouseOver$.unsubscribe();
    this.mouseOut$.unsubscribe();
  }

  render(): React.ReactElement<DropDown> {
    const { children, className, overlay, ...props } = this.props;
    return (
      <StyledDropDown ref={this.dropDownReference} className={classNames(className)} {...props}>
        {children}
        <div hidden={this.state.isHidden} className={classNames('drop-down-container')}>
          <div className="arrow-up" />
          {overlay}
        </div>
      </StyledDropDown>
    );
  }
}

export default DropDown;
