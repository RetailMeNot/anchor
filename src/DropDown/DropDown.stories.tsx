// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
// VENDOR
import styled from 'styled-components';
// COMPONENTS
import DropDown from './DropDown.component';
import { Menu, Item } from '../Menu';
import { List, Item as ListItem } from '../List';
// THEME

const StyledStory = styled.div`
  padding: 2rem 5rem;
`;

const MockList = () => (
  <List>
    <ListItem>Item 1</ListItem>
    <ListItem>Item 1</ListItem>
    <ListItem>Item 1</ListItem>
    <ListItem>Item 1</ListItem>
    <ListItem>Item 1</ListItem>
  </List>
);

storiesOf('Components/DropDown', module)
  .add('Default', () => (
      <StyledStory>
        <p>DropDown 1</p>
          <DropDown overlay={<MockList />}>
            <a>DropDown Link <br/> big</a>
          </DropDown>
        <p>DropDown 2</p>
        <DropDown overlay={<MockList />}>
          <a>DropDown Link</a>
        </DropDown>
        <p>Menu</p>
        <Menu>
            <DropDown overlay={<MockList />}>
              <Item path="/">Categories</Item>
            </DropDown>
            <DropDown overlay={<MockList />}>
              <Item path="/">Stores</Item>
            </DropDown>
        </Menu>
      </StyledStory>
  ));
