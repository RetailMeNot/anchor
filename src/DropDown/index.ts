export { DropDown } from './DropDown.component';
