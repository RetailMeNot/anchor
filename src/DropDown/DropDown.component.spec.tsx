// REACT
import * as React from 'react';
import * as renderer from 'react-test-renderer';
// COMPONENT
import DropDown from './DropDown.component';
// ENZYME
import { shallow, mount } from 'enzyme';

// TEST SETUP
const subject = <DropDown overlay={<div>1</div>} />;
const wrapper = mount(subject);
const component = shallow(subject);

describe('Component: DropDown', () => {
    it('should be defined', () => {
        /* tslint:disable */
        console.log(subject);
        console.log(wrapper.debug());
        console.log(component.debug());

        /* tslint:enable */
        expect(subject).toBeDefined();
        expect(wrapper).toBeDefined();
        expect(component).toBeDefined();
        const tree = renderer.create(subject).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
