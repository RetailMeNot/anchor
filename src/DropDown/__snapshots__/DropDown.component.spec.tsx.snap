// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: DropDown should render in the DOM 1`] = `
.c3 {
  position: absolute;
  box-sizing: border-box;
  width: 0.625rem;
  height: 0.625rem;
  border: light;
  background: white;
  box-shadow: 0 0 0.5rem 0 rgba(0,0,0,0.2);
  z-index: -1;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  top: calc(-0.625rem / 2);
  left: calc(50% - 0.625rem / 2);
}

.c0 {
  box-sizing: border-box;
  position: relative;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 1rem;
  min-width: 1rem;
  padding-bottom: 0.75rem;
  margin-bottom: -0.75rem;
}

.c1 {
  position: absolute;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  z-index: 1;
  min-width: 10rem;
  border: light;
  border-radius: base;
  box-shadow: 0 0 0.5rem 0 rgba(0,0,0,0.2);
  visibility: hidden;
  left: 0;
  top: 0;
  -webkit-transform: translate3d(0px,0px,0);
  -ms-transform: translate3d(0px,0px,0);
  transform: translate3d(0px,0px,0);
}

.c2 {
  box-sizing: border-box;
  border-radius: inherit;
  width: 100%;
  z-index: 2;
  background: white;
}

<div
  className="anchor-drop-down c0"
  spacing="0.75rem"
>
  <div
    className="anchor-drop-down-container c1"
    height={null}
    width={null}
  >
    <div
      className="c2"
    >
      <div>
        1
      </div>
    </div>
    <div
      className="c3"
    />
  </div>
</div>
`;
